the-collab-lab/tcl-11-smart-shopping-list
# Feature(Issue #10): Record item next purchase date in the Firestore database

## Description
This Pull Request (PR) adds a new feature that allows an estimated next purchase date to be computed at the time an item is marked as purchased. Additionally, the newly computed `nextPurchaseInterval` is saved into the Firestore database.

To calculate the `nextPurchaseInterval` it was impartive to: 
- Update the `Listener` component to pass `lastEstimate`, `latestInterval`, `numberOfPurchases`, `nextPurchaseInterval` to `Item` component
- Update `Item` component to `calculateEstimate`, `lastEstimate`, `latestInterval`, `numberOfPurchases` and change them to firestore
- Update the firebase import in `Listener` component
- Update `Add Item` component with new fields `lastEstimate`, `latestInterval`, `numberOfPurchases` and added them to firestore

## Related Issue

Closes Issue #10

## Acceptance Criteria

- [x] The script at `/src/lib/estimates.js` should be used to calculate the next estimated purchase interval.

## Type of Changes


|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

<img width="1293" alt="Before Image" src="https://user-images.githubusercontent.com/32943879/92284880-02c10080-eed1-11ea-8ad3-03eccf3c87a6.png">

### After
<img width="1284" alt="After Image" src="https://user-images.githubusercontent.com/32943879/92284807-de652400-eed0-11ea-9cc4-073ded151406.png">

## Testing Steps / QA Criteria
 
- From your terminal, run git checkout `kl-vh-next-purchase-date-estimator`
- Once on the `kl-vh-next-purchase-date-estimator` branch, run npm start to load the server and start the browser.
- Visit the <Home/> component at http://localhost:3000.
- Create a new shopping list, click the 'Create a New Shopping List' button to be redirected to the <List /> component view. 
- Click the 'Add Item' button' to be redirected to the <AddItem /> component view and add any item to the shopping list.
- After the page reloads and the new item is displayed click the 'check box' to mark the item as purchased. 
- Visit the [Firestore console ](https://console.firebase.google.com/u/1/project/tcl-11-smart-shopping-list/firestore) to access the newly created token shopping list.
- Within the token shopping list view the document details for the newly added item.
- In order to emulate a previous purchase, edit the item `lastPurchaseDate` field and update the date to a past date.
- Visit the <List /> component view at http://localhost:3000/list.
- After the page loads, click the 'checkbox' to mark the item as purchased for a second time. 
- Repeat the process to update the `lastPurchaseDate` field in the Firestore console and return to the browser to mark the item purchased for a total of three times. 
- In order for the `lastEstimate` field to update, an item must be recorded as purchased at least three times.
- Go back to [Firestore console](https://console.firebase.google.com/u/1/project/tcl-11-smart-shopping-list/firestore), now the selected item should have a new `nextPurchaseInterval` value.
